{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "7Gkte68aIhTy"
   },
   "source": [
    "Welcome!\n",
    "\n",
    "So glad you made it and thanks for checking out HyenaDNA! We're \\**very*\\* excited to get this in the hands of researchers.\n",
    "\n",
    "Let's walk through how to setup a simple HyenaDNA model to:\n",
    "\n",
    "*    Train on the GenomicBenchmarks dataset from scratch or fine-tune (sequence level classification tasks)\n",
    "*    Load from HuggingFace :)\n",
    "*    Do inference on up to 1M tokens!\n",
    "\n",
    "--\n",
    "\n",
    "After this tutorial, you can check out more advanced training (more configurable and optimized for speed) through the HyenaDNA repo, which uses Pytorch Lightning and Hydra for managing all the experimental configs and settings.\n",
    "\n",
    "Luckily, the code for classification here is 95% the same as the pretraining code!\n",
    "\n",
    "- [arxiv](https://arxiv.org/abs/2306.15794)  \n",
    "- [blog](https://hazyresearch.stanford.edu/blog/2023-06-29-hyena-dna)\n",
    "- [github](https://github.com/HazyResearch/hyena-dna)  \n",
    "- [huggingface](https://huggingface.co/LongSafari) (pretrained weights)  \n",
    "\n",
    "Good luck and enjoy!!!\n",
    "\n",
    "\\\\\n",
    "Credit: much of the code is forked and extended from [S4](https://github.com/HazyResearch/state-spaces) and [Safari](https://github.com/HazyResearch/safari)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 1000
    },
    "id": "JWUc70H7Iiwd",
    "outputId": "3e03483b-4ccc-4add-e412-a72bfc749f3a"
   },
   "outputs": [],
   "source": [
    "#@title Installs\n",
    "# !pip install einops\n",
    "# !pip install torchvision\n",
    "# !pip install transformers==4.26.1\n",
    "# !pip install genomic-benchmarks\n",
    "# !pip install OmegaConf\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "id": "-WIH4-WEI4in"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11.7\n",
      "Thu Aug 29 14:00:30 2024       \n",
      "+-----------------------------------------------------------------------------+\n",
      "| NVIDIA-SMI 525.85.12    Driver Version: 525.85.12    CUDA Version: 12.0     |\n",
      "|-------------------------------+----------------------+----------------------+\n",
      "| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |\n",
      "| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |\n",
      "|                               |                      |               MIG M. |\n",
      "|===============================+======================+======================|\n",
      "|   0  NVIDIA A100 80G...  On   | 00000000:CA:00.0 Off |                    0 |\n",
      "| N/A   35C    P0    57W / 300W |    121MiB / 81920MiB |      0%      Default |\n",
      "|                               |                      |             Disabled |\n",
      "+-------------------------------+----------------------+----------------------+\n",
      "                                                                               \n",
      "+-----------------------------------------------------------------------------+\n",
      "| Processes:                                                                  |\n",
      "|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |\n",
      "|        ID   ID                                                   Usage      |\n",
      "|=============================================================================|\n",
      "|  No running processes found                                                 |\n",
      "+-----------------------------------------------------------------------------+\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#@title Imports\n",
    "# for HyenaDNA specifically\n",
    "import torch\n",
    "import math\n",
    "import torch\n",
    "import torch.nn as nn\n",
    "import torch.nn.functional as F\n",
    "from functools import partial\n",
    "from einops import rearrange\n",
    "from typing import Optional\n",
    "from functools import partial\n",
    "from torch import Tensor\n",
    "from torchvision.ops import StochasticDepth\n",
    "from collections import namedtuple\n",
    "print(torch.version.cuda)\n",
    "import sys\n",
    "sys.executable.split('/')[-3]\n",
    "!nvidia-smi\n",
    "torch.cuda.device_count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#my imports\n",
    "import ftplib\n",
    "import os\n",
    "import pandas as pd\n",
    "import youtokentome as yttm\n",
    "from functools import partial\n",
    "import gzip\n",
    "from BCBio import GFF \n",
    "import numpy as np\n",
    "import smart_open \n",
    "import gensim \n",
    "from sklearn.model_selection import train_test_split \n",
    "import matplotlib.pyplot as plt \n",
    "import sqlite3\n",
    "# import sqlalchemy \n",
    "from matplotlib import cm\n",
    "from tables import * \n",
    "os.environ['PYTORCH_CUDA_ALLOC_CONF'] = \"\"\n",
    "os.environ['CUDA_LAUNCH_BLOCKING']='1'\n",
    "import datetime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'2024-08-29'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "str(datetime.date.today())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "nu3UYdFqKzTs"
   },
   "source": [
    "# HyenaDNA\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "id": "pWJZc3volbL6"
   },
   "outputs": [],
   "source": [
    "#@title Hyena layer\n",
    "\n",
    "\n",
    "def fftconv(u, k, D):\n",
    "    \"\"\"\n",
    "    We apply a convolution through the fourier domain (from the Convolution Theorem)\n",
    "\n",
    "    \"\"\"\n",
    "    seqlen = u.shape[-1]\n",
    "    fft_size = 2 * seqlen\n",
    "\n",
    "    k_f = torch.fft.rfft(k, n=fft_size) / fft_size\n",
    "    u_f = torch.fft.rfft(u.to(dtype=k.dtype), n=fft_size)\n",
    "\n",
    "    if len(u.shape) > 3: k_f = k_f.unsqueeze(1)\n",
    "    y = torch.fft.irfft(u_f * k_f, n=fft_size, norm='forward')[..., :seqlen]\n",
    "\n",
    "    out = y + u * D.unsqueeze(-1)\n",
    "    return out.to(dtype=u.dtype)\n",
    "\n",
    "\n",
    "@torch.jit.script\n",
    "def mul_sum(q, y):\n",
    "    return (q * y).sum(dim=1)\n",
    "\n",
    "class OptimModule(nn.Module):\n",
    "    \"\"\" Interface for Module that allows registering buffers/parameters with configurable optimizer hyperparameters \"\"\"\n",
    "\n",
    "    def register(self, name, tensor, lr=None, wd=0.0):\n",
    "        \"\"\"Register a tensor with a configurable learning rate and 0 weight decay\"\"\"\n",
    "\n",
    "        if lr == 0.0:\n",
    "            self.register_buffer(name, tensor)\n",
    "        else:\n",
    "            self.register_parameter(name, nn.Parameter(tensor))\n",
    "\n",
    "            optim = {}\n",
    "            if lr is not None: optim[\"lr\"] = lr\n",
    "            if wd is not None: optim[\"weight_decay\"] = wd\n",
    "            setattr(getattr(self, name), \"_optim\", optim)\n",
    "\n",
    "\n",
    "class Sin(nn.Module):\n",
    "    \"\"\"The Sin activation function for the Hyena Filter function.\"\"\"\n",
    "    def __init__(self, dim, w=10, train_freq=True):\n",
    "        super().__init__()\n",
    "        self.freq = nn.Parameter(w * torch.ones(1, dim)) if train_freq else w * torch.ones(1, dim)\n",
    "\n",
    "    def forward(self, x):\n",
    "        return torch.sin(self.freq * x)\n",
    "\n",
    "\n",
    "class PositionalEmbedding(OptimModule):\n",
    "    def __init__(self, emb_dim: int, seq_len: int, lr_pos_emb: float=1e-5, **kwargs):\n",
    "        \"\"\"Complex exponential positional embeddings for Hyena filters.\"\"\"\n",
    "        super().__init__()\n",
    "\n",
    "        self.seq_len = seq_len\n",
    "        # The time embedding fed to the filteres is normalized so that t_f = 1\n",
    "        t = torch.linspace(0, 1, self.seq_len)[None, :, None] # 1, L, 1\n",
    "\n",
    "        if emb_dim > 1:\n",
    "            bands = (emb_dim - 1) // 2\n",
    "        # To compute the right embeddings we use the \"proper\" linspace\n",
    "        t_rescaled = torch.linspace(0, seq_len - 1, seq_len)[None, :, None]\n",
    "        w = 2 * math.pi * t_rescaled / seq_len # 1, L, 1\n",
    "\n",
    "        f = torch.linspace(1e-4, bands - 1, bands)[None, None]\n",
    "        z = torch.exp(-1j * f * w)\n",
    "        z = torch.cat([t, z.real, z.imag], dim=-1)\n",
    "        self.register(\"z\", z, lr=lr_pos_emb)\n",
    "        self.register(\"t\", t, lr=0.0)\n",
    "\n",
    "    def forward(self, L):\n",
    "        return self.z[:, :L], self.t[:, :L]\n",
    "\n",
    "\n",
    "class ExponentialModulation(OptimModule):\n",
    "    \"\"\"The window function applied to the output of the (MLP) filter function.\"\"\"\n",
    "    def __init__(\n",
    "        self,\n",
    "        d_model,\n",
    "        fast_decay_pct=0.3,\n",
    "        slow_decay_pct=1.5,\n",
    "        target=1e-2,\n",
    "        modulation_lr=0.0,\n",
    "        modulate: bool=True,\n",
    "        shift: float = 0.05,\n",
    "        **kwargs\n",
    "    ):\n",
    "        super().__init__()\n",
    "        self.modulate = modulate\n",
    "        self.shift = shift\n",
    "        max_decay = math.log(target) / fast_decay_pct\n",
    "        min_decay = math.log(target) / slow_decay_pct\n",
    "        deltas = torch.linspace(min_decay, max_decay, d_model)[None, None]\n",
    "        self.register(\"deltas\", deltas, lr=modulation_lr)\n",
    "\n",
    "    def forward(self, t, x):\n",
    "        if self.modulate:\n",
    "            decay = torch.exp(-t * self.deltas.abs())\n",
    "            x = x * (decay + self.shift)\n",
    "        return x\n",
    "\n",
    "\n",
    "class HyenaFilter(OptimModule):\n",
    "    def __init__(\n",
    "            self,\n",
    "            d_model,\n",
    "            emb_dim=3, # dim of input to MLP, augments with positional encoding\n",
    "            order=16, # width of the implicit MLP\n",
    "            fused_fft_conv=False,\n",
    "            seq_len=1024,\n",
    "            lr=1e-3,\n",
    "            lr_pos_emb=1e-5,\n",
    "            dropout=0.0,\n",
    "            w=1, # frequency of periodic activations\n",
    "            wd=0, # weight decay of kernel parameters\n",
    "            bias=True,\n",
    "            num_inner_mlps=2,\n",
    "            normalized=False,\n",
    "            **kwargs\n",
    "        ):\n",
    "        \"\"\"\n",
    "        Implicit long filter with modulation.\n",
    "\n",
    "        Args:\n",
    "            d_model: number of channels in the input\n",
    "            emb_dim: dimension of the positional encoding (`emb_dim` - 1) // 2 is the number of bands\n",
    "            order: width of the FFN\n",
    "            num_inner_mlps: number of inner linear layers inside filter MLP\n",
    "\n",
    "        Note:\n",
    "            filter_dropout is not implemented\n",
    "        \"\"\"\n",
    "        super().__init__()\n",
    "\n",
    "        self.d_model = d_model\n",
    "        self.use_bias = bias\n",
    "        self.fused_fft_conv = fused_fft_conv\n",
    "        self.bias = nn.Parameter(torch.randn(self.d_model))\n",
    "        self.dropout = nn.Dropout(dropout)\n",
    "\n",
    "        act = Sin(dim=order, w=w)\n",
    "        self.emb_dim = emb_dim\n",
    "        assert emb_dim % 2 != 0 and emb_dim >= 3, \"emb_dim must be odd and greater or equal to 3 (time, sine and cosine)\"\n",
    "        self.seq_len = seq_len\n",
    "\n",
    "        self.pos_emb = PositionalEmbedding(emb_dim, seq_len, lr_pos_emb)\n",
    "\n",
    "        self.implicit_filter = nn.Sequential(\n",
    "            nn.Linear(emb_dim, order),\n",
    "            act,\n",
    "        )\n",
    "        for i in range(num_inner_mlps):\n",
    "            self.implicit_filter.append(nn.Linear(order, order))\n",
    "            self.implicit_filter.append(act)\n",
    "\n",
    "        self.implicit_filter.append(nn.Linear(order, d_model, bias=False))\n",
    "\n",
    "        self.modulation = ExponentialModulation(d_model, **kwargs)\n",
    "\n",
    "        self.normalized = normalized\n",
    "        for c in self.implicit_filter.children():\n",
    "            for name, v in c.state_dict().items():\n",
    "                optim = {\"weight_decay\": wd, \"lr\": lr}\n",
    "                setattr(getattr(c, name), \"_optim\", optim)\n",
    "\n",
    "    def filter(self, L, *args, **kwargs):\n",
    "        z, t = self.pos_emb(L)\n",
    "        h = self.implicit_filter(z)\n",
    "        h = self.modulation(t, h)\n",
    "        return h\n",
    "\n",
    "    def forward(self, x, L, k=None, bias=None, *args, **kwargs):\n",
    "        if k is None: k = self.filter(L)\n",
    "\n",
    "        # Ensure compatibility with filters that return a tuple\n",
    "        k = k[0] if type(k) is tuple else k\n",
    "        # print(\"shape of x and k:\", x.shape, k.shape)\n",
    "        y = fftconv(x, k, bias)\n",
    "        # print(\"shape of y:\", y.shape)\n",
    "        return y\n",
    "\n",
    "\n",
    "class HyenaOperator(nn.Module):\n",
    "    def __init__(\n",
    "            self,\n",
    "            d_model,\n",
    "            l_max,\n",
    "            order=2,\n",
    "            filter_order=64,\n",
    "            dropout=0.0,\n",
    "            filter_dropout=0.0,\n",
    "            **filter_args,\n",
    "        ):\n",
    "        r\"\"\"\n",
    "        Hyena operator described in the paper https://arxiv.org/pdf/2302.10866.pdf\n",
    "\n",
    "        Args:\n",
    "            d_model (int): Dimension of the input and output embeddings (width of the layer)\n",
    "            l_max: (int): Maximum input sequence length. Defaults to None\n",
    "            order: (int): Depth of the Hyena recurrence. Defaults to 2\n",
    "            dropout: (float): Dropout probability. Defaults to 0.0\n",
    "            filter_dropout: (float): Dropout probability for the filter. Defaults to 0.0\n",
    "        \"\"\"\n",
    "        super().__init__()\n",
    "\n",
    "        self.d_model = d_model\n",
    "        self.l_max = l_max\n",
    "        self.order = order\n",
    "        inner_width = d_model * (order + 1)\n",
    "        self.dropout = nn.Dropout(dropout)\n",
    "        self.in_proj = nn.Linear(d_model, inner_width)\n",
    "        self.out_proj = nn.Linear(d_model, d_model)\n",
    "\n",
    "        self.short_filter = nn.Conv1d(\n",
    "            inner_width,\n",
    "            inner_width,\n",
    "            3,\n",
    "            padding=2,\n",
    "            groups=inner_width\n",
    "        )\n",
    "        self.filter_fn = HyenaFilter(\n",
    "            d_model * (order - 1),\n",
    "            order=filter_order,\n",
    "            seq_len=l_max,\n",
    "            channels=1,\n",
    "            dropout=filter_dropout,\n",
    "            **filter_args\n",
    "        )\n",
    "\n",
    "    def forward(self, u, *args, **kwargs):\n",
    "        l = u.size(-2)\n",
    "        l_filter = min(l, self.l_max)\n",
    "        # print(\"shape of u:\",u.shape)\n",
    "        u = self.in_proj(u)\n",
    "        u = rearrange(u, 'b l d -> b d l')\n",
    "\n",
    "        uc = self.short_filter(u)[...,:l_filter]\n",
    "        *x, v = uc.split(self.d_model, dim=1)\n",
    "\n",
    "        k = self.filter_fn.filter(l_filter)[0]\n",
    "        k = rearrange(k, 'l (o d) -> o d l', o=self.order - 1)\n",
    "        bias = rearrange(self.filter_fn.bias, '(o d) -> o d', o=self.order - 1)\n",
    "\n",
    "        for o, x_i in enumerate(reversed(x[1:])):\n",
    "            v = self.dropout(v * x_i)\n",
    "            v = self.filter_fn(v, l_filter, k=k[o], bias=bias[o])\n",
    "\n",
    "        y = rearrange(v * x[0], 'b d l -> b l d')\n",
    "\n",
    "        y = self.out_proj(y)\n",
    "        # print(\"y shape (operator):\", y.shape)\n",
    "        return y\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "id": "2I7LjzVvr9fp"
   },
   "outputs": [],
   "source": [
    "#@title Self-Attention (alternative)\n",
    "\n",
    "\"\"\"\n",
    "If you'd like to try the HyenaDNA model using attention instead, you can. ie,\n",
    "use a regular decoder only Transformer.\n",
    "\n",
    "Borrowed from the FlashAttention library by Tri Dao.\n",
    "\"\"\"\n",
    "\n",
    "class SelfAttention(nn.Module):\n",
    "    \"\"\"Implement the scaled dot product attention with softmax.\n",
    "    Arguments\n",
    "    ---------\n",
    "        softmax_scale: The temperature to use for the softmax attention.\n",
    "                      (default: 1/sqrt(d_keys) where d_keys is computed at\n",
    "                      runtime)\n",
    "        attention_dropout: The dropout rate to apply to the attention\n",
    "                           (default: 0.0)\n",
    "    \"\"\"\n",
    "    def __init__(self, causal=False, softmax_scale=None, attention_dropout=0.0):\n",
    "        super().__init__()\n",
    "        self.causal = causal\n",
    "        self.softmax_scale = softmax_scale\n",
    "        self.dropout_p = attention_dropout\n",
    "\n",
    "    def forward(self, qkv, causal=None, key_padding_mask=None):\n",
    "        \"\"\"Implements the multihead softmax attention.\n",
    "        Arguments\n",
    "        ---------\n",
    "            qkv: The tensor containing the query, key, and value. (B, S, 3, H, D)\n",
    "            causal: if passed, will override self.causal\n",
    "            key_padding_mask: boolean mask to apply to the attention weights. True means to keep,\n",
    "                False means to mask out. (B, S)\n",
    "        \"\"\"\n",
    "        batch_size, seqlen = qkv.shape[0], qkv.shape[1]\n",
    "        causal = self.causal if causal is None else causal\n",
    "        q, k, v = qkv.unbind(dim=2)\n",
    "        softmax_scale = self.softmax_scale or 1.0 / math.sqrt(q.shape[-1])\n",
    "        scores = torch.einsum('bthd,bshd->bhts', q, k * softmax_scale)\n",
    "        if key_padding_mask is not None:\n",
    "            padding_mask = torch.full((batch_size, seqlen), -10000.0, dtype=scores.dtype,\n",
    "                                      device=scores.device)\n",
    "            padding_mask.masked_fill_(key_padding_mask, 0.0)\n",
    "            scores = scores + rearrange(padding_mask, 'b s -> b 1 1 s')\n",
    "        if causal:\n",
    "            # \"triu_tril_cuda_template\" not implemented for 'BFloat16'\n",
    "            # So we have to construct the mask in float\n",
    "            causal_mask = torch.triu(torch.full((seqlen, seqlen), -10000.0, device=scores.device), 1)\n",
    "            scores = scores + causal_mask.to(dtype=scores.dtype)\n",
    "        attention = torch.softmax(scores, dim=-1, dtype=v.dtype)\n",
    "        attention_drop = F.dropout(attention, self.dropout_p if self.training else 0.0)\n",
    "        output = torch.einsum('bhts,bshd->bthd', attention_drop, v)\n",
    "        return output\n",
    "\n",
    "class MHA(nn.Module):\n",
    "    \"\"\"Multi-head self-attention and cross-attention\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, embed_dim, num_heads, bias=True, dropout=0.0,\n",
    "                 softmax_scale=None, causal=False, layer_idx=None, dwconv=False,return_residual=False,device=None, dtype=None) -> None:\n",
    "        \"\"\"\n",
    "            return_residual: whether to return the input x along with the output. This is for\n",
    "                performance reason: for post-norm architecture, returning the input allows us\n",
    "                to fuse the backward of nn.Linear with the residual connection.\n",
    "        \"\"\"\n",
    "        factory_kwargs = {'device': device, 'dtype': dtype}\n",
    "        super().__init__()\n",
    "        self.embed_dim = embed_dim\n",
    "        self.causal = causal\n",
    "        self.layer_idx = layer_idx\n",
    "        self.dwconv = dwconv\n",
    "        self.return_residual = return_residual\n",
    "\n",
    "        self.num_heads = num_heads\n",
    "        assert self.embed_dim % num_heads == 0, \"self.kdim must be divisible by num_heads\"\n",
    "        self.head_dim = self.embed_dim // num_heads\n",
    "\n",
    "        linear_cls = nn.Linear\n",
    "        linear_resid_cls = LinearResidual\n",
    "        inner_attn_cls =  SelfAttention\n",
    "\n",
    "        if not self.return_residual:\n",
    "            self.Wqkv = linear_cls(embed_dim, 3 * embed_dim, bias=bias, **factory_kwargs)\n",
    "        else:\n",
    "            self.Wqkv = linear_resid_cls(embed_dim, 3 * embed_dim, bias=bias, **factory_kwargs)\n",
    "        if self.dwconv:\n",
    "            self.dwconv_qkv = nn.Conv1d(3 * embed_dim, 3 * embed_dim, kernel_size=3, padding=2,\n",
    "                                        groups=3 * embed_dim)\n",
    "\n",
    "        self.inner_attn = inner_attn_cls(causal=causal, softmax_scale=softmax_scale,\n",
    "                                         attention_dropout=dropout)\n",
    "\n",
    "        # output projection always have the bias (for now)\n",
    "        self.out_proj = linear_cls(embed_dim, embed_dim, **factory_kwargs)\n",
    "\n",
    "    def forward(self, x, key_padding_mask=None, **kwargs):\n",
    "        \"\"\"\n",
    "        Arguments:\n",
    "            x: (batch, seqlen, hidden_dim) (where hidden_dim = num heads * head dim) if\n",
    "                cu_seqlens is None and max_seqlen is None, else (total, hidden_dim) where total\n",
    "                is the is the sum of the sequence lengths in the batch.\n",
    "            cu_seqlens: (batch_size + 1,), dtype torch.int32. The cumulative sequence lengths\n",
    "                of the sequences in the batch, used to index into x. Only applicable when using\n",
    "                FlashAttention.\n",
    "            max_seqlen: int. Maximum sequence length in the batch.\n",
    "            key_padding_mask: boolean mask, True means to keep, False means to mask out.\n",
    "                (batch, seqlen). Only applicable when not using FlashAttention.\n",
    "            mixer_subset: for cross-attention only. If not None, will take a subset of x\n",
    "                before applying the query projection. Useful for e.g., ViT where we only care\n",
    "                about the CLS token in the last layer.\n",
    "            inference_params: for generation. Adapted from Megatron-LM (and Apex)\n",
    "            https://github.com/NVIDIA/apex/blob/3ff1a10f72ec07067c4e44759442329804ac5162/apex/transformer/testing/standalone_transformer_lm.py#L470\n",
    "        \"\"\"\n",
    "\n",
    "        kwargs = ({'key_padding_mask': key_padding_mask, **kwargs})\n",
    "\n",
    "        if not self.return_residual:\n",
    "            qkv = self.Wqkv(x)\n",
    "        else:\n",
    "            qkv, x = self.Wqkv(x)\n",
    "        if self.dwconv:\n",
    "            qkv = rearrange(self.dwconv_qkv(rearrange(qkv, 'b s d -> b d s'))[..., :-2],\n",
    "                            'b d s -> b s d').contiguous()\n",
    "        qkv = rearrange(qkv, '... (three h d) -> ... three h d', three=3, d=self.head_dim)\n",
    "\n",
    "        context = self.inner_attn(qkv, **kwargs)\n",
    "\n",
    "        out = self.out_proj(rearrange(context, '... h d -> ... (h d)'))\n",
    "        return out if not self.return_residual else (out, x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "id": "dgkZggOetkbR"
   },
   "outputs": [],
   "source": [
    "#@title MLP layer\n",
    "\n",
    "\"\"\"\n",
    "The MLP layer after the mixer layer (HyenaOperator).\n",
    "\"\"\"\n",
    "\n",
    "class Mlp(nn.Module):\n",
    "\n",
    "    def __init__(self, in_features, hidden_features=None, out_features=None, activation=F.gelu,\n",
    "                 return_residual=False, device=None, dtype=None):\n",
    "        \"\"\"\n",
    "        From https://github.com/HazyResearch/flash-attention/blob/main/flash_attn/modules/mlp.py\n",
    "        \"\"\"\n",
    "        factory_kwargs = {'device': device, 'dtype': dtype}\n",
    "        super().__init__()\n",
    "        out_features = out_features or in_features\n",
    "        hidden_features = hidden_features or in_features\n",
    "        self.return_residual = return_residual\n",
    "        self.fc1 = nn.Linear(in_features, hidden_features, **factory_kwargs)\n",
    "        self.activation = activation\n",
    "        self.fc2 = nn.Linear(hidden_features, out_features, **factory_kwargs)\n",
    "\n",
    "    def forward(self, x):\n",
    "        y = self.fc1(x)\n",
    "        y = self.activation(y)\n",
    "        y = self.fc2(y)\n",
    "        return y if not self.return_residual else (y, x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "id": "xAC-fIFFstei"
   },
   "outputs": [],
   "source": [
    "#@title Block layer (Hyena + MLP layers)\n",
    "\n",
    "\"\"\"\n",
    "A block consists of a Mixer layer (Hyena or attention), and a MLP layer.\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "class LinearResidual(nn.Linear):\n",
    "    \"\"\"Wrap nn.Linear to return the residual as well. For compatibility with FusedDense.\n",
    "    \"\"\"\n",
    "\n",
    "    def forward(self, input: torch.Tensor) -> torch.Tensor:\n",
    "        return super().forward(input), input\n",
    "\n",
    "class Block(nn.Module):\n",
    "\n",
    "    def __init__(self, dim, mixer_cls=None, mlp_cls=None, norm_cls=nn.LayerNorm,\n",
    "                 dropout_cls=nn.Dropout, prenorm=True, resid_dropout1=0., resid_dropout2=0.,\n",
    "                 drop_path1=0., drop_path2=0.,\n",
    "                 return_residual=False,\n",
    "                 residual_in_fp32=False):\n",
    "        \"\"\"\n",
    "        From https://github.com/HazyResearch/flash-attention/blob/main/flash_attn/modules/block.py\n",
    "        For prenorm=True, this Block has a slightly different structure compared to a regular\n",
    "        prenorm Transformer block.\n",
    "        The standard block is: LN -> MHA -> Dropout -> Add -> LN -> MLP -> Dropout -> Add.\n",
    "        [Ref: https://arxiv.org/abs/2002.04745]\n",
    "        Here we have: Dropout -> Add -> LN -> MHA -> Dropout -> Add -> LN -> MLP, returning both\n",
    "        the hidden_states (output of the MLP) and the residual.\n",
    "        This is for performance reasons, as we can fuse the dropout, add and LayerNorm.\n",
    "        The residual needs to be provided (except for the very first block).\n",
    "        For prenorm=False, this Block has the same structure as a regular postnorm Transformer\n",
    "        block: MHA -> Dropout -> Add -> LN -> MLP -> Dropout -> Add -> LN.\n",
    "        return_residual: whether each of the sub-layers (mixer and mlp) will return the residual.\n",
    "        This is for performance reason: for post-norm architecture, returning the input allows us\n",
    "        to fuse the backward of nn.Linear with the residual connection.\n",
    "        \"\"\"\n",
    "        super().__init__()\n",
    "        self.prenorm = prenorm\n",
    "        self.return_residual = return_residual\n",
    "        self.residual_in_fp32 = residual_in_fp32\n",
    "        if self.residual_in_fp32:\n",
    "            assert self.prenorm, 'residual_in_fp32 is only compatible with prenorm=True'\n",
    "        if mixer_cls is None:\n",
    "            mixer_cls = partial(MHA, num_heads=dim // 64)\n",
    "        if mlp_cls is None:\n",
    "            mlp_cls = partial(Mlp, hidden_features=4 * dim)\n",
    "        self.mixer = mixer_cls()\n",
    "        self.dropout1 = dropout_cls(resid_dropout1)\n",
    "        self.drop_path1 = StochasticDepth(drop_path1, mode='row')\n",
    "        self.norm1 = norm_cls(dim)\n",
    "        self.mlp = mlp_cls(dim)\n",
    "        if not isinstance(self.mlp, nn.Identity):\n",
    "            self.dropout2 = dropout_cls(resid_dropout2)\n",
    "            self.drop_path2 = StochasticDepth(drop_path2, mode='row')\n",
    "            self.norm2 = norm_cls(dim)\n",
    "\n",
    "    def forward(self, hidden_states, residual = None,\n",
    "                mixer_subset=None, mixer_kwargs=None):\n",
    "        r\"\"\"Pass the input through the encoder layer.\n",
    "        Args:\n",
    "            hidden_states: the sequence to the encoder layer (required).\n",
    "            residual: if postnorm, residual=None, If prenorm, hidden_states = Attn/MLP(LN(residual))\n",
    "            mixer_subset: for cross-attention only. If not None, will take a subset of x\n",
    "                before applying the query projection. Useful for e.g., ViT where we only care\n",
    "                about the CLS token in the last layer.\n",
    "        \"\"\"\n",
    "        if self.prenorm:\n",
    "            dropped = self.drop_path1(self.dropout1(hidden_states))\n",
    "            residual = (dropped + residual) if residual is not None else dropped\n",
    "            hidden_states = self.norm1(residual.to(dtype=self.norm1.weight.dtype))\n",
    "            if self.residual_in_fp32:\n",
    "                residual = residual.to(torch.float32)\n",
    "            if mixer_kwargs is None:\n",
    "                mixer_kwargs = {}\n",
    "            if mixer_subset is not None:\n",
    "                mixer_kwargs['mixer_subset'] = mixer_subset\n",
    "            hidden_states = self.mixer(hidden_states, **mixer_kwargs)\n",
    "            if mixer_subset is not None:\n",
    "                residual = residual[:, mixer_subset]\n",
    "            if not isinstance(self.mlp, nn.Identity):\n",
    "                dropped = self.drop_path2(self.dropout2(hidden_states))\n",
    "                residual = (dropped + residual) if residual is not None else dropped\n",
    "                hidden_states = self.norm2(residual.to(dtype=self.norm2.weight.dtype))\n",
    "                if self.residual_in_fp32:\n",
    "                    residual = residual.to(torch.float32)\n",
    "\n",
    "                hidden_states = self.mlp(hidden_states)\n",
    "            return hidden_states, residual\n",
    "        else:\n",
    "            assert residual is None\n",
    "            mixer_out = self.mixer(\n",
    "                hidden_states, **(mixer_kwargs if mixer_kwargs is not None else {})\n",
    "            )\n",
    "            if self.return_residual:  # mixer out is actually a pair here\n",
    "                mixer_out, hidden_states = mixer_out\n",
    "\n",
    "            hidden_states = self.norm1((self.drop_path1(self.dropout1(mixer_out))\n",
    "                                        + hidden_states).to(dtype=self.norm1.weight.dtype))\n",
    "\n",
    "            if not isinstance(self.mlp, nn.Identity):\n",
    "                mlp_out = self.mlp(hidden_states)\n",
    "                if self.return_residual:  # mlp out is actually a pair here\n",
    "                    mlp_out, hidden_states = mlp_out\n",
    "\n",
    "                hidden_states = self.norm2((self.drop_path2(self.dropout2(mlp_out))\n",
    "                                            + hidden_states).to(dtype=self.norm2.weight.dtype))\n",
    "\n",
    "            return hidden_states\n",
    "\n",
    "def create_mixer_cls(layer=None,\n",
    "                     attn_layer_idx=None, attn_cfg=None, layer_idx=None,\n",
    "                     device=None, dtype=None):\n",
    "    factory_kwargs = {'device': device, 'dtype': dtype}\n",
    "    if attn_layer_idx is not None and layer_idx in attn_layer_idx:\n",
    "        causal = True if attn_cfg is None else attn_cfg.pop('causal', True)\n",
    "\n",
    "        mha_cls = MHA\n",
    "\n",
    "        mixer_cls = partial(mha_cls, causal=causal, layer_idx=layer_idx,\n",
    "                            **(attn_cfg if attn_cfg is not None else {}),**factory_kwargs)\n",
    "    else:\n",
    "        # mixer_cls = instantiate(registry.layer, layer, partial=True, layer_idx=layer_idx, **factory_kwargs)\n",
    "\n",
    "        mixer_cls = partial(HyenaOperator, **layer)\n",
    "\n",
    "    return mixer_cls\n",
    "\n",
    "def create_mlp_cls(d_model, d_inner=None, device=None, dtype=None):\n",
    "    factory_kwargs = {'device': device, 'dtype': dtype}\n",
    "    inner_dim = d_inner if d_inner is not None else 4 * d_model\n",
    "\n",
    "    mlp_cls = partial(Mlp, hidden_features=inner_dim,\n",
    "                          activation=partial(F.gelu, approximate='tanh'), **factory_kwargs)\n",
    "\n",
    "    return mlp_cls\n",
    "\n",
    "\n",
    "def create_block(d_model, d_inner=None,\n",
    "                 layer=None, attn_layer_idx=None,\n",
    "                 attn_cfg=None, layer_norm_epsilon=1e-5,\n",
    "                 resid_dropout1=0.0, resid_dropout2=0.0, residual_in_fp32=False,\n",
    "                 layer_idx=None,\n",
    "                 device=None, dtype=None):\n",
    "    factory_kwargs = {'device': device, 'dtype': dtype}\n",
    "    mixer_cls = create_mixer_cls(layer=layer,\n",
    "                                 attn_layer_idx=attn_layer_idx,\n",
    "                                 attn_cfg=attn_cfg, layer_idx=layer_idx,\n",
    "                                 **factory_kwargs)\n",
    "    mlp_cls = create_mlp_cls(d_model, d_inner=d_inner,\n",
    "                             **factory_kwargs)\n",
    "    norm_cls = partial(nn.LayerNorm, eps=layer_norm_epsilon, **factory_kwargs)\n",
    "    block = Block(d_model, mixer_cls, mlp_cls, norm_cls=norm_cls,\n",
    "                  prenorm=True, resid_dropout1=resid_dropout1, resid_dropout2=resid_dropout2,residual_in_fp32=residual_in_fp32)\n",
    "    block.layer_idx = layer_idx\n",
    "    return block\n",
    "\n",
    "\n",
    "# https://github.com/huggingface/transformers/blob/c28d04e9e252a1a099944e325685f14d242ecdcd/src/transformers/models/gpt2/modeling_gpt2.py#L454\n",
    "def _init_weights(module, n_layer, initializer_range=0.02, rescale_prenorm_residual=True,\n",
    "                  glu_act=False):\n",
    "    if isinstance(module, nn.Linear):\n",
    "        nn.init.normal_(module.weight, std=initializer_range)\n",
    "        if module.bias is not None:\n",
    "            nn.init.zeros_(module.bias)\n",
    "    elif isinstance(module, nn.Embedding):\n",
    "        nn.init.normal_(module.weight, std=initializer_range)\n",
    "\n",
    "    if rescale_prenorm_residual:\n",
    "        # Reinitialize selected weights subject to the OpenAI GPT-2 Paper Scheme:\n",
    "        #   > A modified initialization which accounts for the accumulation on the residual path with model depth. Scale\n",
    "        #   > the weights of residual layers at initialization by a factor of 1/âˆšN where N is the # of residual layers.\n",
    "        #   >   -- GPT-2 :: https://openai.com/blog/better-language-models/\n",
    "        #\n",
    "        # Reference (Megatron-LM): https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/model/gpt_model.py\n",
    "        for name, p in module.named_parameters():\n",
    "            if name in [\"out_proj.weight\", \"fc2.weight\"]:\n",
    "                # Special Scaled Initialization --> There are 2 Layer Norms per Transformer Block\n",
    "                nn.init.normal_(p, mean=0.0, std=initializer_range / math.sqrt(2 * n_layer))\n",
    "            # If using GLU activation for now, we scale the std by 2\n",
    "            elif name in [\"output_linear.0.weight\"]:\n",
    "                # Special Scaled Initialization --> There are 2 Layer Norms per Transformer Block\n",
    "                if not glu_act:\n",
    "                    nn.init.normal_(p, mean=0.0, std=initializer_range / math.sqrt(2 * n_layer))\n",
    "                else:\n",
    "                    out_features = p.shape[0]\n",
    "                    # Multiplying the first half of the matrix by 2 since sigmoid scales it down by 0.5\n",
    "                    # on average.\n",
    "                    nn.init.normal_(p[:out_features // 2], mean=0.0, std=initializer_range / math.sqrt(2 * n_layer) * 2)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "id": "OJHCegJvtASc"
   },
   "outputs": [],
   "source": [
    "#@title Backbone model (stack of blocks)\n",
    "\n",
    "\"\"\"\n",
    "A backbone model consists of a stack of blocks. If you use attention, then\n",
    "positional embeddings are included. When using Hyena, then the pos emb\n",
    "revert to doing nothing.\n",
    "\"\"\"\n",
    "\n",
    "class GPT2Embeddings(nn.Module):\n",
    "\n",
    "    def __init__(self, embed_dim, vocab_size, max_position_embeddings, padding_idx=None,\n",
    "                 word_embed_proj_dim=None, device=None, dtype=None):\n",
    "        \"\"\"\n",
    "            If max_position_embeddings <= 0, there's no position embeddings\n",
    "            If word_embe_proj_dim is not None (e.g., OPT-350m), we embed to that dimension\n",
    "                the project up to embed_dim\n",
    "        \"\"\"\n",
    "        factory_kwargs = {'device': device, 'dtype': dtype}\n",
    "        super().__init__()\n",
    "        if word_embed_proj_dim is None:\n",
    "            self.word_embeddings = nn.Embedding(vocab_size, embed_dim, padding_idx=padding_idx,\n",
    "                                                **factory_kwargs)\n",
    "            self.project_in = None\n",
    "        else:\n",
    "            self.word_embeddings = nn.Embedding(vocab_size, word_embed_proj_dim,\n",
    "                                                padding_idx=padding_idx, **factory_kwargs)\n",
    "            self.project_in = nn.Linear(word_embed_proj_dim, embed_dim, bias=False,\n",
    "                                        **factory_kwargs)\n",
    "        self.max_position_embeddings = max_position_embeddings\n",
    "        if self.max_position_embeddings > 0:\n",
    "            self.position_embeddings = nn.Embedding(max_position_embeddings, embed_dim,\n",
    "                                                    **factory_kwargs)\n",
    "\n",
    "    def forward(self, input_ids, position_ids=None):\n",
    "        \"\"\"\n",
    "            input_ids: (batch, seqlen)\n",
    "            position_ids: (batch, seqlen)\n",
    "        \"\"\"\n",
    "        batch_size, seqlen = input_ids.shape\n",
    "        embeddings = self.word_embeddings(input_ids)\n",
    "        if self.project_in is not None:\n",
    "            embeddings = self.project_in(embeddings)\n",
    "        if self.max_position_embeddings > 0:\n",
    "            if position_ids is None:\n",
    "                position_ids = torch.arange(seqlen, dtype=torch.long, device=input_ids.device)\n",
    "            position_embeddings = self.position_embeddings(position_ids)\n",
    "            embeddings = embeddings + position_embeddings\n",
    "        return embeddings\n",
    "\n",
    "class LMBackbone(nn.Module):\n",
    "\n",
    "    def __init__(self, d_model: int, n_layer: int, d_inner: int, vocab_size: int,\n",
    "                 process_group=None, layer=None,\n",
    "                 attn_layer_idx=None, attn_cfg=None, max_position_embeddings=0,\n",
    "                 resid_dropout: float = 0.0, embed_dropout: float = 0.1,\n",
    "                 layer_norm_epsilon: float = 1e-5, initializer_cfg=None,residual_in_fp32=False,\n",
    "                 device=None, dtype=None, **kwargs) -> None:\n",
    "        factory_kwargs = {'device': device, 'dtype': dtype}\n",
    "        super().__init__()\n",
    "        self.process_group = process_group\n",
    "        self.residual_in_fp32 = residual_in_fp32\n",
    "        # note max_position_embeddings is 0 for Hyena, and therefore isn't used\n",
    "        self.embeddings = GPT2Embeddings(d_model, vocab_size, max_position_embeddings,\n",
    "                                             **factory_kwargs)\n",
    "\n",
    "        self.layers = nn.ModuleList([create_block(\n",
    "            d_model, d_inner=d_inner,\n",
    "            layer=layer, attn_layer_idx=attn_layer_idx,\n",
    "            attn_cfg=attn_cfg, layer_norm_epsilon=layer_norm_epsilon,\n",
    "            resid_dropout1=embed_dropout if i == 0 else resid_dropout,\n",
    "            resid_dropout2=resid_dropout, residual_in_fp32=residual_in_fp32,layer_idx=i,\n",
    "            **factory_kwargs,\n",
    "        ) for i in range(n_layer)])\n",
    "\n",
    "        self.drop_f = nn.Dropout(resid_dropout)\n",
    "        self.ln_f = nn.LayerNorm(d_model, eps=layer_norm_epsilon, **factory_kwargs)\n",
    "\n",
    "        self.apply(partial(_init_weights, n_layer=n_layer,\n",
    "                           **(initializer_cfg if initializer_cfg is not None else {})))\n",
    "\n",
    "    def forward(self, input_ids, position_ids=None):\n",
    "        hidden_states = self.embeddings(input_ids, position_ids=position_ids,)\n",
    "        residual = None\n",
    "\n",
    "        for layer in self.layers:\n",
    "            hidden_states, residual = layer(hidden_states, residual)\n",
    "\n",
    "        dropped = self.drop_f(hidden_states)\n",
    "        residual = (dropped + residual) if residual is not None else dropped\n",
    "        hidden_states = self.ln_f(residual.to(dtype=self.ln_f.weight.dtype))\n",
    "\n",
    "        return hidden_states"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "id": "CR0wFk-CawPB"
   },
   "outputs": [],
   "source": [
    "#@title Decoder head layer\n",
    "\n",
    "\"\"\"\n",
    "A simple decoder head (using MLP) to predict a sequence level classification.\n",
    "You have the option to average across all the tokens in a sequence or using the\n",
    "\"last\" token to classify.  At least, those 2 worked best for us, but we provide\n",
    "other \"modes\" as well.\n",
    "\n",
    "We only need this for classification.  Otherwise we'll use the hidden\n",
    "states of the backbone as embeddings.\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "class SequenceDecoder(nn.Module):\n",
    "    def __init__(\n",
    "        self, d_model, d_output=None, l_output=None, use_lengths=False, mode=\"last\"\n",
    "    ):\n",
    "        super().__init__()\n",
    "\n",
    "        self.output_transform = nn.Identity() if d_output is None else nn.Linear(d_model, d_output)\n",
    "\n",
    "        if l_output is None:\n",
    "            self.l_output = None\n",
    "            self.squeeze = False\n",
    "        elif l_output == 0:\n",
    "            # Equivalent to getting an output of length 1 and then squeezing\n",
    "            self.l_output = 1\n",
    "            self.squeeze = True\n",
    "        else:\n",
    "            assert l_output > 0\n",
    "            self.l_output = l_output\n",
    "            self.squeeze = False\n",
    "\n",
    "        self.use_lengths = use_lengths\n",
    "        self.mode = mode\n",
    "\n",
    "        if mode == 'ragged':\n",
    "            assert not use_lengths\n",
    "\n",
    "    def forward(self, x, state=None, lengths=None, l_output=None):\n",
    "        \"\"\"\n",
    "        x: (n_batch, l_seq, d_model)\n",
    "        Returns: (n_batch, l_output, d_output)\n",
    "        \"\"\"\n",
    "\n",
    "        if self.l_output is None:\n",
    "            if l_output is not None:\n",
    "                assert isinstance(l_output, int)  # Override by pass in\n",
    "            else:\n",
    "                # Grab entire output\n",
    "                l_output = x.size(-2)\n",
    "            squeeze = False\n",
    "        else:\n",
    "            l_output = self.l_output\n",
    "            squeeze = self.squeeze\n",
    "\n",
    "        if self.mode == \"last\":\n",
    "            restrict = lambda x: x[..., -l_output:, :]\n",
    "        elif self.mode == \"first\":\n",
    "            restrict = lambda x: x[..., :l_output, :]\n",
    "        elif self.mode == \"pool\":\n",
    "            restrict = lambda x: (\n",
    "                torch.cumsum(x, dim=-2)\n",
    "                / torch.arange(\n",
    "                    1, 1 + x.size(-2), device=x.device, dtype=x.dtype\n",
    "                ).unsqueeze(-1)\n",
    "            )[..., -l_output:, :]\n",
    "\n",
    "            def restrict(x):\n",
    "                L = x.size(-2)\n",
    "                s = x.sum(dim=-2, keepdim=True)\n",
    "                if l_output > 1:\n",
    "                    c = torch.cumsum(x[..., -(l_output - 1) :, :].flip(-2), dim=-2)\n",
    "                    c = F.pad(c, (0, 0, 1, 0))\n",
    "                    s = s - c  # (B, l_output, D)\n",
    "                    s = s.flip(-2)\n",
    "                denom = torch.arange(\n",
    "                    L - l_output + 1, L + 1, dtype=x.dtype, device=x.device\n",
    "                )\n",
    "                s = s / denom\n",
    "                return s\n",
    "\n",
    "        elif self.mode == \"sum\":\n",
    "            restrict = lambda x: torch.cumsum(x, dim=-2)[..., -l_output:, :]\n",
    "            # TODO use same restrict function as pool case\n",
    "        elif self.mode == 'ragged':\n",
    "            assert lengths is not None, \"lengths must be provided for ragged mode\"\n",
    "            # remove any additional padding (beyond max length of any sequence in the batch)\n",
    "            restrict = lambda x: x[..., : max(lengths), :]\n",
    "        else:\n",
    "            raise NotImplementedError(\n",
    "                \"Mode must be ['last' | 'first' | 'pool' | 'sum']\"\n",
    "            )\n",
    "\n",
    "        # Restrict to actual length of sequence\n",
    "        if self.use_lengths:\n",
    "            assert lengths is not None\n",
    "            x = torch.stack(\n",
    "                [\n",
    "                    restrict(out[..., :length, :])\n",
    "                    for out, length in zip(torch.unbind(x, dim=0), lengths)\n",
    "                ],\n",
    "                dim=0,\n",
    "            )\n",
    "        else:\n",
    "            x = restrict(x)\n",
    "\n",
    "        if squeeze:\n",
    "            assert x.size(-2) == 1\n",
    "            x = x.squeeze(-2)\n",
    "\n",
    "        x = self.output_transform(x)\n",
    "\n",
    "        return x\n",
    "\n",
    "    def step(self, x, state=None):\n",
    "        # Ignore all length logic\n",
    "        return self.output_transform(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "id": "ZM3g4ImZtKNu"
   },
   "outputs": [],
   "source": [
    "#@title Model (backbone + head)\n",
    "\n",
    "\"\"\"\n",
    "Putting it all together, the model consists of a backbone model\n",
    "and a decoder head (you can turn off head for embeddings only too).\n",
    "\n",
    "Here we use a simple head to do multi-classification, but\n",
    "can also swap the head to do next token prediction too.  We defer to the main\n",
    "HyenaDNA for that code, since pretraining with next token prediction isn't quite\n",
    "feasible on colab.\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "class HyenaDNAModel(nn.Module):\n",
    "\n",
    "    def __init__(self, d_model: int, n_layer: int, d_inner: int, vocab_size: int,\n",
    "                 layer=None, attn_layer_idx=None, attn_cfg=None, max_position_embeddings=0,\n",
    "                 resid_dropout: float = 0.0, embed_dropout: float = 0.1,\n",
    "                 layer_norm_epsilon: float = 1e-5, initializer_cfg=None,residual_in_fp32=False,\n",
    "                 pad_vocab_size_multiple: int = 1, use_head=False, n_classes: int = 2,\n",
    "                 device=None, dtype=None, **kwargs) -> None:\n",
    "        factory_kwargs = {'device': device, 'dtype': dtype}\n",
    "        super().__init__()\n",
    "        if vocab_size % pad_vocab_size_multiple != 0:\n",
    "            vocab_size += pad_vocab_size_multiple - (vocab_size % pad_vocab_size_multiple)\n",
    "\n",
    "        self.use_head = use_head\n",
    "\n",
    "        # check if layer (config) has d_model (HF code differs from main Safari code)\n",
    "        if 'd_model' not in layer:\n",
    "            layer['d_model'] = d_model\n",
    "\n",
    "        self.backbone = LMBackbone(\n",
    "            d_model=d_model, n_layer=n_layer, d_inner=d_inner, vocab_size=vocab_size,\n",
    "            layer=layer, attn_layer_idx=attn_layer_idx, attn_cfg=attn_cfg,\n",
    "            max_position_embeddings=max_position_embeddings,\n",
    "            resid_dropout=resid_dropout, embed_dropout=embed_dropout,\n",
    "            layer_norm_epsilon=layer_norm_epsilon,\n",
    "            initializer_cfg=initializer_cfg, residual_in_fp32=residual_in_fp32,\n",
    "            **factory_kwargs, **kwargs\n",
    "        )\n",
    "\n",
    "        # we only need a head if doing classification, otherwise we'll use the\n",
    "        # hidden states as embeddings\n",
    "        if self.use_head:\n",
    "            self.head = SequenceDecoder(d_model=d_model, d_output=n_classes, l_output=0, mode='pool')\n",
    "\n",
    "        # Initialize weights and apply final processing\n",
    "        self.apply(partial(_init_weights, n_layer=n_layer,\n",
    "                           **(initializer_cfg if initializer_cfg is not None else {})))\n",
    "\n",
    "        # if self.use_head:\n",
    "        #     self.tie_weights()\n",
    "\n",
    "    # def tie_weights(self):\n",
    "    #     self.head.weight = self.backbone.embeddings.word_embeddings.weight\n",
    "\n",
    "    def forward(self, input_ids, position_ids=None, state=None): # state for the repo interface\n",
    "        hidden_states = self.backbone(input_ids, position_ids=position_ids)\n",
    "\n",
    "        if self.use_head:\n",
    "            return self.head(hidden_states)\n",
    "        else:\n",
    "            return hidden_states"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "id": "3gGiI05HX-AU"
   },
   "outputs": [],
   "source": [
    "#@title Huggingface Pretrained Wrapper\n",
    "# for Huggingface integration, we use a wrapper class around the model\n",
    "# to load weights\n",
    "import json\n",
    "import os\n",
    "import subprocess\n",
    "import transformers\n",
    "from transformers import PreTrainedModel, AutoModelForCausalLM, PretrainedConfig\n",
    "import re\n",
    "\n",
    "def inject_substring(orig_str):\n",
    "    \"\"\"Hack to handle matching keys between models trained with and without\n",
    "    gradient checkpointing.\"\"\"\n",
    "\n",
    "    # modify for mixer keys\n",
    "    pattern = r\"\\.mixer\"\n",
    "    injection = \".mixer.layer\"\n",
    "\n",
    "    modified_string = re.sub(pattern, injection, orig_str)\n",
    "\n",
    "    # modify for mlp keys\n",
    "    pattern = r\"\\.mlp\"\n",
    "    injection = \".mlp.layer\"\n",
    "\n",
    "    modified_string = re.sub(pattern, injection, modified_string)\n",
    "\n",
    "    return modified_string\n",
    "\n",
    "def load_weights(scratch_dict, pretrained_dict, checkpointing=False):\n",
    "    \"\"\"Loads pretrained (backbone only) weights into the scratch state dict.\n",
    "\n",
    "    scratch_dict: dict, a state dict from a newly initialized HyenaDNA model\n",
    "    pretrained_dict: dict, a state dict from the pretrained ckpt\n",
    "    checkpointing: bool, whether the gradient checkpoint flag was used in the\n",
    "    pretrained model ckpt. This slightly changes state dict keys, so we patch\n",
    "    that if used.\n",
    "\n",
    "    return:\n",
    "    dict, a state dict with the pretrained weights loaded (head is scratch)\n",
    "\n",
    "    # loop thru state dict of scratch\n",
    "    # find the corresponding weights in the loaded model, and set it\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    # need to do some state dict \"surgery\"\n",
    "    for key, value in scratch_dict.items():\n",
    "        if 'backbone' in key:\n",
    "            # the state dicts differ by one prefix, '.model', so we add that\n",
    "            key_loaded = 'model.' + key\n",
    "            # breakpoint()\n",
    "            # need to add an extra \".layer\" in key\n",
    "            if checkpointing:\n",
    "                key_loaded = inject_substring(key_loaded)\n",
    "            try:\n",
    "                scratch_dict[key] = pretrained_dict[key_loaded]\n",
    "            except:\n",
    "                raise Exception('key mismatch in the state dicts!')\n",
    "\n",
    "    # scratch_dict has been updated\n",
    "    return scratch_dict\n",
    "\n",
    "class HyenaDNAPreTrainedModel(PreTrainedModel):\n",
    "    \"\"\"\n",
    "    An abstract class to handle weights initialization and a simple interface for downloading and loading pretrained\n",
    "    models.\n",
    "    \"\"\"\n",
    "    base_model_prefix = \"hyenadna\"\n",
    "\n",
    "    def __init__(self, config):\n",
    "        pass\n",
    "\n",
    "    def forward(self, input_ids, **kwargs):\n",
    "        return self.model(input_ids, **kwargs)\n",
    "\n",
    "    @classmethod\n",
    "    def from_pretrained(cls,\n",
    "                        path,\n",
    "                        model_name,\n",
    "                        download=False,\n",
    "                        config=None,\n",
    "                        device='cpu',\n",
    "                        use_head=False,\n",
    "                        n_classes=2,\n",
    "                      ):\n",
    "        # first check if it is a local path\n",
    "        pretrained_model_name_or_path = os.path.join(path, model_name)\n",
    "        if os.path.isdir(pretrained_model_name_or_path) and download == False:\n",
    "            if config is None:\n",
    "                config = json.load(open(os.path.join(pretrained_model_name_or_path, 'config.json')))\n",
    "        else:\n",
    "            hf_url = f'https://huggingface.co/LongSafari/{model_name}'\n",
    "\n",
    "            subprocess.run(f'rm -rf {pretrained_model_name_or_path}', shell=True)\n",
    "            command = f'mkdir -p {path} && cd {path} && git lfs install && git clone {hf_url}'\n",
    "            subprocess.run(command, shell=True)\n",
    "\n",
    "            if config is None:\n",
    "                config = json.load(open(os.path.join(pretrained_model_name_or_path, 'config.json')))\n",
    "\n",
    "        scratch_model = HyenaDNAModel(**config, use_head=use_head, n_classes=n_classes)  # the new model format\n",
    "        loaded_ckpt = torch.load(\n",
    "            os.path.join(pretrained_model_name_or_path, 'weights.ckpt'),\n",
    "            map_location=torch.device(device)\n",
    "        )\n",
    "\n",
    "        # need to load weights slightly different if using gradient checkpointing\n",
    "        if config.get(\"checkpoint_mixer\", False):\n",
    "            checkpointing = config[\"checkpoint_mixer\"] == True or config[\"checkpoint_mixer\"] == True\n",
    "        else:\n",
    "            checkpointing = False\n",
    "\n",
    "        # grab state dict from both and load weights\n",
    "        state_dict = load_weights(scratch_model.state_dict(), loaded_ckpt['state_dict'], checkpointing=checkpointing)\n",
    "\n",
    "        # scratch model has now been updated\n",
    "        scratch_model.load_state_dict(state_dict)\n",
    "        print(\"Loaded pretrained weights ok!\")\n",
    "        return scratch_model\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "qsRya6hEryfg"
   },
   "source": [
    "# Data pipeline\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "id": "vsL1Afn5flo0"
   },
   "outputs": [],
   "source": [
    "#@title Dataloader\n",
    "from torch.utils.data import Dataset, DataLoader\n",
    "from torch.nn.utils.rnn import pad_sequence\n",
    "from torch import nn, optim\n",
    "import torch\n",
    "from tables import *\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "from sklearn.model_selection import train_test_split\n",
    "import time\n",
    "import math\n",
    "\n",
    "class UHGGDatasetH5(Dataset):\n",
    "    def __init__(self, df, class_dict, path_to_db, pad_length=1069919, label_col = 'family', pad_token = 0):\n",
    "        super().__init__()\n",
    "        self.df = df\n",
    "        self.idx_dict = dict(zip(np.arange(len(df)), list(df.index)))\n",
    "        self.label_col = label_col\n",
    "        self.class_dict = class_dict #dict(zip(df[label_col].unique(), np.arange(len(df[label_col].unique()))))\n",
    "        # print(len(df[label_col].unique()))\n",
    "        # print(len(df), len(self.class_dict.keys()))\n",
    "        self.path = path_to_db\n",
    "        self.file = open_file(path_to_db, 'r')\n",
    "        self.data = self.file.root.data\n",
    "        self.pad_length = pad_length\n",
    "        self.pad_token = pad_token\n",
    "    def __len__(self):\n",
    "        return(len(self.df))\n",
    "    def pad_to_length(self, tkns):\n",
    "        if(len(tkns)<self.pad_length):\n",
    "            tkns += [0] * (self.pad_length-len(tkns))\n",
    "        else:\n",
    "            tkns=tkns[:self.pad_length]\n",
    "        return tkns\n",
    "    def __getitem__(self, idx):\n",
    "        index = self.idx_dict[idx]\n",
    "        tokens = self.data['Tknz%d' % index].read()\n",
    "        x = self.pad_to_length(tokens)\n",
    "        x = torch.tensor(x)\n",
    "        species_rep = self.df[self.label_col][index]\n",
    "        y = self.class_dict[species_rep]\n",
    "        y = torch.tensor(y).long()\n",
    "        return (x.cuda(), y.cuda())\n",
    "\n",
    "# def Collater(object):\n",
    "#     def __init__(self, pad_token):\n",
    "#         self.pad_token = pad_token\n",
    "#     def __call__(self, data):\n",
    "#         x, y = data\n",
    "#         x = pad_sequence(x, padding_value = pad_token)\n",
    "#         return x, y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "id": "U2_Uw320e6mk"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.4526615606034683\n",
      "CPU times: user 308 ms, sys: 35.6 ms, total: 344 ms\n",
      "Wall time: 353 ms\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "64992"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "batch_size = 1\n",
    "label_col = \"family\"\n",
    "#np.random.seed(42)\n",
    "#torch.manual_seed(42)\n",
    "path = '/gpfs/data/johnsonslab/nlp-genomics/almeida-2020/'\n",
    "path_to_db = os.path.join(path, 'tokenizations_8192k_poc.h5')\n",
    "all_species = pd.read_csv(os.path.join(path, \"all_species_clades.csv\"))\n",
    "family_counts = all_species['family'].value_counts()\n",
    "n_classes = 4\n",
    "print(sum(family_counts[:n_classes])/sum(family_counts[:2]))\n",
    "samples = all_species[all_species['family'].isin(family_counts.index[:n_classes])]\n",
    "train_set, val_set = train_test_split(samples, test_size=0.2)\n",
    "val_set, test_set = train_test_split(val_set, test_size=0.5)\n",
    "class_dict = dict(zip(samples['family'].unique(), np.arange(len(samples['family'].unique())))) ###necessary to define class dict to ensure label fidelity\n",
    "\n",
    "# train_loader = DataLoader(UHGGDatasetH5(train_set, class_dict, path_to_db, label_col=label_col,\n",
    "#                                         pad_token = 0, pad_length = 1069919),\n",
    "#                         batch_size=batch_size, shuffle=True)#, collate_fn=Collater(vocab_size))#, num_workers=4)\n",
    "# val_loader = DataLoader(UHGGDatasetH5(val_set, class_dict, path_to_db, label_col=label_col, \n",
    "#                                       pad_token=0, pad_length = 1069919),\n",
    "#                         batch_size=batch_size, shuffle=True)#, collate_fn=Collater(vocab_size))#, num_workers=4)\n",
    "# test_loader = DataLoader(UHGGDatasetH5(test_set, class_dict, path_to_db, label_col = label_col, \n",
    "#                                        pad_token=0, pad_length = 1069919),\n",
    "#                         batch_size=batch_size, shuffle=True)#, collate_fn=Collater(vocab_size))#, num_workers=4)\n",
    "# print(train_loader.dataset.class_dict, val_loader.dataset.class_dict, test_loader.dataset.class_dict)\n",
    "family_counts[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "family\n",
       "Lachnospiraceae       64992\n",
       "Bacteroidaceae        34367\n",
       "Ruminococcaceae       26022\n",
       "Acutalibacteraceae    18954\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "family_counts[:4]\n",
    "# class_0 = samples[samples['family'] == family_counts.index[0]]\n",
    "# train_0, val_0 = split_by_clade(class_0, 0.8, column='genus')\n",
    "# val_0, test_0 = split_by_clade(val_0, 0.5, column='genus')\n",
    "\n",
    "# len(test_0['genus'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# class_1 = samples[samples['family'] == family_counts.index[1]]\n",
    "# train_1, val_1 = split_by_clade(class_1, 0.8, column='genus')\n",
    "# val_1, test_1 = split_by_clade(val_0, 0.5, column='genus')\n",
    "\n",
    "# test_1.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tqdm import tqdm\n",
    "zeroes = 0\n",
    "ones = 0\n",
    "#lachno = 0, bacter = 1\n",
    "lachno_lens = []\n",
    "bacter_lens = []\n",
    "# print(train_loader.dataset.class_dict)\n",
    "    # if label.item() == 0:\n",
    "    #     zeroes+=1\n",
    "    # elif label.item()==1:\n",
    "    #     ones+=1\n",
    "\n",
    "# print(len(train_loader) - (zeroes + 1))\n",
    "\n",
    "# for (data, label) in (test_loader):\n",
    "#     if label.item() == 0:\n",
    "#         zeroes+=1\n",
    "    # elif label.item()==1:\n",
    "#         ones+=1\n",
    "#     print(data.shape, label)\n",
    "#     if (zeroes !=0) and (ones != 0):\n",
    "#         print (float(ones)/zeroes)\n",
    "# print((len(test_loader) + len(train_loader)) - (zeroes + 1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# y = np.array([0, 1, 2, 2])\n",
    "\n",
    "# pred = np.array([[.7, .1, .05], [.1, .8, .05], [.6, .2, .9],[.1,  .1, .2]])\n",
    "# print(pred.shape)\n",
    "# binary_y = np.zeros(pred.shape)\n",
    "# binary_y[np.arange(y.size), y] = 1\n",
    "# print(binary_y)\n",
    "# # pred = np.max(pred, axis=1)\n",
    "# print(pred)\n",
    "# auc = roc_auc_score(binary_y, pred)\n",
    "# auc\n",
    "# #pred = np.array([0.1, 0.4, 0.35, 0.5])\n",
    "# # fpr, tpr, thresholds = roc_curve(y, pred, pos_label=2)\n",
    "# # auc(fpr, tpr)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "False\n"
     ]
    }
   ],
   "source": [
    "# a1 = np.ndarray([3, 4])\n",
    "# # a2 = np.ndarray([1, 4])\n",
    "# a3 = np.concatenate((a1, a2), axis=0)\n",
    "# a3.shape\n",
    "a1 = np.ndarray([1,4])\n",
    "print(type(a1)==int)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "5vMJu0wQWUBN"
   },
   "source": [
    "### Training! (and fine-tuning)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "id": "dMNRAa4gKJOK"
   },
   "outputs": [],
   "source": [
    "import torch.optim as optim\n",
    "import time\n",
    "from sklearn.metrics import auc, roc_curve, roc_auc_score, classification_report, confusion_matrix\n",
    "\"\"\"\n",
    "We provide simple training code for the GenomicBenchmark datasets.\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "def train(model, device, train_loader, optimizer, epoch, loss_fn, day, log_interval=10000):\n",
    "    \"\"\"Training loop.\"\"\"\n",
    "    model.train()\n",
    "    start_time = time.time()\n",
    "    for batch_idx, (data, target) in enumerate(train_loader):\n",
    "        data, target = data.to(device), target.to(device)\n",
    "        optimizer.zero_grad()\n",
    "        output = model(data)\n",
    "        #print(\"input shape:\", data.shape, \"target shape: \", target.shape)\n",
    "\n",
    "        #print(\"successfully processed data\")\n",
    "        #print(target)\n",
    "        #print(\"output shape:\", output.shape, \"target shape: \", target.shape)\n",
    "        loss = loss_fn(output, target)\n",
    "        #print(\"loss computed\")\n",
    "\n",
    "        loss.backward()\n",
    "        #print(\"loss backward complete\")\n",
    "        optimizer.step()\n",
    "        #print(\"full training step completed\")\n",
    "        if batch_idx % log_interval == 0:\n",
    "            print('Train Epoch: {} [{}/{} ({:.0f}%)]\\tLoss: {:.6f} | Elapsed time: {}'.format(\n",
    "                epoch, batch_idx * len(data), len(train_loader),\n",
    "                100. * batch_idx / len(train_loader), loss.item(), time.time() - start_time))\n",
    "        torch.cuda.empty_cache()\n",
    "    print(\"Epoch {} complete, elapsed time: {}\".format(epoch, time.time() - start_time))\n",
    "    torch.save({'model_state_dict':model.state_dict(), 'optimizer_state_dict':optimizer.state_dict(),'loss':loss}, \n",
    "               '/gpfs/scratch/jic286/JohnsonLab/hyena-model-checkpoints/hyena-bpe-4_families-epoch_{}-{}.pt'.format(epoch, day))\n",
    "\n",
    "def test(model, device, test_loader, loss_fn):\n",
    "    \"\"\"Test loop.\"\"\"\n",
    "    model.eval()\n",
    "    test_loss = 0\n",
    "    correct = 0\n",
    "    preds = 0\n",
    "    truths = 0\n",
    "    with torch.no_grad():\n",
    "        for data, target in (test_loader):\n",
    "            data, target = data.to(device), target.to(device)\n",
    "            output = model(data)\n",
    "            if type(preds) == int:\n",
    "                preds = output.cpu().numpy()\n",
    "            else:\n",
    "                preds = np.concatenate((preds, output.cpu().numpy()), axis=0)\n",
    "            if type(truths) == int:\n",
    "                truth = target.cpu().numpy()\n",
    "                truths = np.zeros(output.shape)\n",
    "                truths[np.arange(truth.size), truth] = 1\n",
    "            else:\n",
    "                truth = target.cpu().numpy()\n",
    "                binary_y = np.zeros(output.shape)\n",
    "                binary_y[np.arange(truth.size), truth] = 1\n",
    "                truths = np.concatenate((truths, binary_y), axis=0)\n",
    "            test_loss += loss_fn(output, target).item()  # sum up batch loss\n",
    "            pred = output.argmax(dim=1, keepdim=True)  # get the index of the max log-probability\n",
    "            correct += pred.eq(target.view_as(pred)).sum().item()\n",
    "    print(truths.shape, preds.shape)\n",
    "    auc = roc_auc_score(truths, preds)\n",
    "    y_test = np.argmax(truths, axis=1)\n",
    "    y_pred = np.argmax(preds, axis=1)\n",
    "    test_loss /= len(test_loader)\n",
    "    print(f\"Confusion matrix: {confusion_matrix(y_test, y_pred)}\\n\") \n",
    "    print('Test set: Average loss: {:.4f}, Accuracy: {}/{} ({:.2f}%), AUC: {}\\n'.format(\n",
    "        test_loss, correct, len(test_loader),\n",
    "        100. * correct / len(test_loader), auc))\n",
    "    print(classification_report(y_test, y_pred))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "id": "z6sguA32WZE3"
   },
   "outputs": [],
   "source": [
    "import json\n",
    "import os\n",
    "import subprocess\n",
    "import transformers\n",
    "from transformers import PreTrainedModel, AutoModelForCausalLM, PretrainedConfig\n",
    "\n",
    "def split_by_clade(sample, split, column='species'):\n",
    "    val_counts = sample[column].value_counts()\n",
    "    # print(sample, split, column, val_counts)\n",
    "\n",
    "    total = 0\n",
    "    idx = 0\n",
    "    for i in range(len(val_counts)):\n",
    "        idx = i\n",
    "        total += val_counts[i]\n",
    "        if float(total)/float(sum(val_counts)) >= split:\n",
    "            break\n",
    "        \n",
    "    train_set = sample[sample[column].isin(val_counts.index[:idx+1])]\n",
    "    test_set = sample[sample[column].isin(val_counts.index[idx+1:])]\n",
    "    return train_set, test_set\n",
    "    \n",
    "def run_train():\n",
    "\n",
    "    '''\n",
    "    Main entry point for training.  Select the dataset name and metadata, as\n",
    "    well as model and training args, and you're off to the genomic races!\n",
    "\n",
    "    ### GenomicBenchmarks Metadata\n",
    "    # there are 8 datasets in this suite, choose 1 at a time, with their corresponding settings\n",
    "    # name                                num_seqs        num_classes     median len    std\n",
    "    # dummy_mouse_enhancers_ensembl       1210            2               2381          984.4\n",
    "    # demo_coding_vs_intergenomic_seqs    100_000         2               200           0\n",
    "    # demo_human_or_worm                  100_000         2               200           0\n",
    "    # human_enhancers_cohn                27791           2               500           0\n",
    "    # human_enhancers_ensembl             154842          2               269           122.6\n",
    "    # human_ensembl_regulatory            289061          3               401           184.3\n",
    "    # human_nontata_promoters             36131           2               251           0\n",
    "    # human_ocr_ensembl                   174756          2               315           108.1\n",
    "\n",
    "    '''\n",
    "    # experiment settings:\n",
    "    num_epochs = 1  # ~100 seems fine\n",
    "    max_length = 500  # max len of sequence of dataset (of what you want)\n",
    "    use_padding = True\n",
    "    dataset_name = 'human_enhancers_cohn'\n",
    "    batch_size = 256\n",
    "    learning_rate = 6e-4  # good default for Hyena\n",
    "    rc_aug = True  # reverse complement augmentation\n",
    "    add_eos = False  # add end of sentence token\n",
    "    weight_decay = 0.1\n",
    "\n",
    "    # for fine-tuning, only the 'tiny' model can fit on colab\n",
    "    pretrained_model_name = 'hyenadna-tiny-1k-seqlen'  # use None if training from scratch\n",
    "\n",
    "    # we need these for the decoder head, if using\n",
    "    use_head = True\n",
    "    n_classes = 2\n",
    "\n",
    "    # you can override with your own backbone config here if you want,\n",
    "    # otherwise we'll load the HF one by default\n",
    "    backbone_cfg = None\n",
    "\n",
    "    device = torch.device(\"cuda:0\" if torch.cuda.is_available() else \"cpu\")\n",
    "    print(\"Using device:\", device)\n",
    "    backbone_cfg = {\n",
    "      \"d_model\": 128,\n",
    "      \"n_layer\": 1,\n",
    "      \"d_inner\": 512,\n",
    "      \"vocab_size\": 8192000,\n",
    "      \"resid_dropout\": 0.0,\n",
    "      \"embed_dropout\": 0.1,\n",
    "      \"fused_mlp\": False,\n",
    "      \"fused_dropout_add_ln\": True,\n",
    "      \"checkpoint_mixer\": True,\n",
    "      \"checkpoint_mlp\": True,\n",
    "      \"residual_in_fp32\": True,\n",
    "      \"pad_vocab_size_multiple\": 8,\n",
    "      \"return_hidden_state\": True,\n",
    "      \"layer\": {\n",
    "        \"_name_\": \"hyena\",\n",
    "        \"emb_dim\": 5,\n",
    "        \"filter_order\": 64,\n",
    "        \"local_order\": 3,\n",
    "        \"l_max\": 1069919,\n",
    "        \"modulate\": True,\n",
    "        \"w\": 10,\n",
    "        \"lr\": 6e-4,\n",
    "        \"wd\": 0.0,\n",
    "        \"lr_pos_emb\": 0.0\n",
    "      }\n",
    "    }\n",
    "    ## Original backbone cfg\n",
    "    # backbone_cfg = {\n",
    "    #   \"d_model\": 256,\n",
    "    #   \"n_layer\": 8,\n",
    "    #   \"d_inner\": 1024,\n",
    "    #   \"vocab_size\": 8192000,\n",
    "    #   \"resid_dropout\": 0.0,\n",
    "    #   \"embed_dropout\": 0.1,\n",
    "    #   \"fused_mlp\": False,\n",
    "    #   \"fused_dropout_add_ln\": True,\n",
    "    #   \"checkpoint_mixer\": True,\n",
    "    #   \"checkpoint_mlp\": True,\n",
    "    #   \"residual_in_fp32\": True,\n",
    "    #   \"pad_vocab_size_multiple\": 8,\n",
    "    #   \"return_hidden_state\": True,\n",
    "    #   \"layer\": {\n",
    "    #     \"_name_\": \"hyena\",\n",
    "    #     \"emb_dim\": 5,\n",
    "    #     \"filter_order\": 64,\n",
    "    #     \"local_order\": 3,\n",
    "    #     \"l_max\": 1069919,\n",
    "    #     \"modulate\": True,\n",
    "    #     \"w\": 10,\n",
    "    #     \"lr\": 6e-4,\n",
    "    #     \"wd\": 0.0,\n",
    "    #     \"lr_pos_emb\": 0.0\n",
    "    #   }\n",
    "    # }\n",
    "    # instantiate the model (pretrained here)\n",
    "    # if pretrained_model_name in ['hyenadna-tiny-1k-seqlen']:\n",
    "    #     # use the pretrained Huggingface wrapper instead\n",
    "    #     model = HyenaDNAPreTrainedModel.from_pretrained(\n",
    "    #         './checkpoints',\n",
    "    #         pretrained_model_name,\n",
    "    #         download=True,\n",
    "    #         config=backbone_cfg,\n",
    "    #         device=device,\n",
    "    #         use_head=use_head,\n",
    "    #         n_classes=n_classes,\n",
    "    #     )\n",
    "\n",
    "    # # from scratch\n",
    "    # else:\n",
    "    model = HyenaDNAModel(**backbone_cfg, use_head=use_head, n_classes=n_classes)\n",
    "    # model = nn.DataParallel(model,device_ids=list(range(torch.cuda.device_count())))\n",
    "    # create tokenizer\n",
    "    # tokenizer = CharacterTokenizer(\n",
    "    #     characters=['A', 'C', 'G', 'T', 'N'],  # add DNA characters, N is uncertain\n",
    "    #     model_max_length=max_length + 2,  # to account for special tokens, like EOS\n",
    "    #     add_special_tokens=False,  # we handle special tokens elsewhere\n",
    "    #     padding_side='left', # since HyenaDNA is causal, we pad on the left\n",
    "    # )\n",
    "\n",
    "    # create datasets\n",
    "    # ds_train = GenomicBenchmarkDataset(\n",
    "    #     max_length = max_length,\n",
    "    #     use_padding = use_padding,\n",
    "    #     split = 'train',\n",
    "    #     tokenizer=tokenizer,\n",
    "    #     dataset_name=dataset_name,\n",
    "    #     rc_aug=rc_aug,\n",
    "    #     add_eos=add_eos,\n",
    "    # )\n",
    "\n",
    "    # ds_test = GenomicBenchmarkDataset(\n",
    "    #     max_length = max_length,\n",
    "    #     use_padding = use_padding,\n",
    "    #     split = 'test',\n",
    "    #     tokenizer=tokenizer,\n",
    "    #     dataset_name=dataset_name,\n",
    "    #     rc_aug=rc_aug,\n",
    "    #     add_eos=add_eos,\n",
    "    # )\n",
    "    batch_size = 1\n",
    "    label_col = \"family\"\n",
    "    #np.random.seed(42)\n",
    "    #torch.manual_seed(42)\n",
    "    path = '/gpfs/data/johnsonslab/nlp-genomics/almeida-2020/'\n",
    "    path_to_db = os.path.join(path, 'tokenizations_8192k_poc.h5')\n",
    "    all_species = pd.read_csv(os.path.join(path, \"all_species_clades.csv\"))\n",
    "    family_counts = all_species['family'].value_counts()\n",
    "    class_buffer = 2\n",
    "    samples = all_species[all_species['family'].isin(family_counts.index[class_buffer:n_classes + class_buffer])]\n",
    "    # train_set, val_set = train_test_split(samples, test_size=0.2)\n",
    "    # val_set, test_set = train_test_split(val_set, test_size=0.5)\n",
    "    train_samples = []\n",
    "    val_samples = []\n",
    "    test_samples = []\n",
    "    # print(family_counts[class_buffer:n_classes + class_buffer])\n",
    "    # print(samples['family'].unique(), family_counts.index[2])\n",
    "    for i in range(class_buffer, n_classes + class_buffer):\n",
    "        train_data, val_data = split_by_clade(samples[samples['family']==family_counts.index[i]], 0.8, column='genus')\n",
    "        train_samples.append(train_data)\n",
    "        val_data, test_data = split_by_clade(val_data, 0.5, column='genus')\n",
    "        val_samples.append(val_data)\n",
    "        test_samples.append(test_data)\n",
    "    train_set = pd.concat(train_samples, axis=0)\n",
    "    val_set = pd.concat(val_samples, axis=0)\n",
    "    test_set = pd.concat(test_samples, axis=0)\n",
    "    class_dict = dict(zip(samples['family'].unique(), np.arange(len(samples['family'].unique())))) ###necessary to define class dict to ensure label fidelity\n",
    "    \n",
    "    train_loader = DataLoader(UHGGDatasetH5(train_set, class_dict, path_to_db, label_col=label_col,\n",
    "                                            pad_token = 0, pad_length = 1069919),\n",
    "                            batch_size=batch_size, shuffle=True)#, collate_fn=Collater(vocab_size))#, num_workers=4)\n",
    "    val_loader = DataLoader(UHGGDatasetH5(val_set, class_dict, path_to_db, label_col=label_col, \n",
    "                                          pad_token=0, pad_length = 1069919),\n",
    "                            batch_size=batch_size, shuffle=True)#, collate_fn=Collater(vocab_size))#, num_workers=4)\n",
    "    test_loader = DataLoader(UHGGDatasetH5(test_set, class_dict, path_to_db, label_col = label_col, \n",
    "                                           pad_token=0, pad_length = 1069919),\n",
    "                            batch_size=batch_size, shuffle=True)#, collate_fn=Collater(vocab_size))#, num_workers=4)\n",
    "    print(\"len(train_loader): \", len(train_loader), \" len(train_loader.dataset): \", len(train_loader.dataset))\n",
    "    # train_loader = DataLoader(ds_train, batch_size=batch_size, shuffle=True)\n",
    "    # test_loader = DataLoader(ds_test, batch_size=batch_size, shuffle=False)\n",
    "\n",
    "    # loss function\n",
    "    loss_fn = nn.CrossEntropyLoss()\n",
    "\n",
    "    # create optimizer\n",
    "    optimizer = optim.AdamW(model.parameters(), lr=learning_rate, weight_decay=weight_decay)\n",
    "\n",
    "    model.to(device)\n",
    "    checkpoint = torch.load('/gpfs/scratch/jic286/JohnsonLab/hyena-model-checkpoints/hyena-bpe-4_families-genus_split-epoch_35-2024-04-11.pt')\n",
    "    model.load_state_dict(checkpoint['model_state_dict'])\n",
    "    optimizer.load_state_dict(checkpoint['optimizer_state_dict'])\n",
    "    day = str(datetime.date.today())\n",
    "    for epoch in range(num_epochs):\n",
    "        #train(model, device, train_loader, optimizer, epoch, loss_fn, day=day)\n",
    "        test(model, device, test_loader, loss_fn)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Thu Aug 29 14:25:48 2024       \n",
      "+-----------------------------------------------------------------------------+\n",
      "| NVIDIA-SMI 525.85.12    Driver Version: 525.85.12    CUDA Version: 12.0     |\n",
      "|-------------------------------+----------------------+----------------------+\n",
      "| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |\n",
      "| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |\n",
      "|                               |                      |               MIG M. |\n",
      "|===============================+======================+======================|\n",
      "|   0  NVIDIA A100 80G...  On   | 00000000:CA:00.0 Off |                    0 |\n",
      "| N/A   35C    P0    76W / 300W |  43463MiB / 81920MiB |      0%      Default |\n",
      "|                               |                      |             Disabled |\n",
      "+-------------------------------+----------------------+----------------------+\n",
      "                                                                               \n",
      "+-----------------------------------------------------------------------------+\n",
      "| Processes:                                                                  |\n",
      "|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |\n",
      "|        ID   ID                                                   Usage      |\n",
      "|=============================================================================|\n",
      "|    0   N/A  N/A   2432166      C   .../hyena-dna-new/bin/python    43342MiB |\n",
      "+-----------------------------------------------------------------------------+\n"
     ]
    }
   ],
   "source": [
    "for rank in range(0):\n",
    "    print(f'{rank}, max memory alloc: {torch.cuda.max_memory_allocated(device=rank)}')\n",
    "!nvidia-smi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 1000
    },
    "id": "GhYYO_NP-Au4",
    "outputId": "1c518a04-1872-4fe9-c5f9-c57150e84cd2",
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using device: cuda:0\n",
      "len(train_loader):  37013  len(train_loader.dataset):  37013\n",
      "(3496, 2) (3496, 2)\n",
      "Confusion matrix: [[1920   53]\n",
      " [ 879  644]]\n",
      "\n",
      "Test set: Average loss: 1.8190, Accuracy: 2564/3496 (73.34%), AUC: 0.6816438365737856\n",
      "\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.69      0.97      0.80      1973\n",
      "           1       0.92      0.42      0.58      1523\n",
      "\n",
      "    accuracy                           0.73      3496\n",
      "   macro avg       0.80      0.70      0.69      3496\n",
      "weighted avg       0.79      0.73      0.71      3496\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# launch it!\n",
    "run_train()  # uncomment to run\n",
    "\n",
    "# Using device: cuda:0\n",
    "# Lachnospiraceae    64992\n",
    "# Bacteroidaceae     34367\n",
    "# Ruminococcaceae    26022\n",
    "# Name: family, dtype: int64\n",
    "# shape of u: torch.Size([1, 1069919, 256])\n",
    "# shape of x and k: torch.Size([1, 256, 1069919]) torch.Size([256, 1069919])\n",
    "# shape of y: torch.Size([1, 256, 1069919])\n",
    "# y shape (operator): torch.Size([1, 1069919, 256])\n",
    "# shape of u: torch.Size([1, 1069919, 256])\n",
    "# shape of x and k: torch.Size([1, 256, 1069919]) torch.Size([256, 1069919])\n",
    "#Loss: 0.669707 | Elapsed time: 1.6526081562042236"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "OMgEPV0J92Xj"
   },
   "source": [
    "#### Inference (450k to 1M tokens)!\n",
    "\n",
    "If all you're interested in is getting embeddings on long DNA sequences\n",
    "(inference), then we can do that right here in Colab!\n",
    "\n",
    "\n",
    "*   We provide an example how to load the weights from Huggingface.\n",
    "*   On the free tier, which uses a\n",
    "T4 GPU w/16GB of memory, we can process 450k tokens / nucleotides.\n",
    "*   For processing 1M tokens, you'll need an A100, which Colab offers as a paid tier.\n",
    "*   (Don't forget to run the entire notebook above too)\n",
    "\n",
    "--\n",
    "\n",
    "To pretrain or fine-tune the 1M long sequence model (8 layers, d_model=256),\n",
    "you'll need 8 A100s 80GB, and all that code is in the main [github repo](https://github.com/HazyResearch/hyena-dna)!\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "id": "-1wq2uwUctPV"
   },
   "outputs": [],
   "source": [
    "#@title Single example\n",
    "import json\n",
    "import os\n",
    "import subprocess\n",
    "import transformers\n",
    "from transformers import PreTrainedModel, AutoModelForCausalLM, PretrainedConfig\n",
    "\n",
    "def inference_single():\n",
    "\n",
    "    '''\n",
    "    this selects which backbone to use, and grabs weights/ config from HF\n",
    "    4 options:\n",
    "      'hyenadna-tiny-1k-seqlen'   # fine-tune on colab ok\n",
    "      'hyenadna-small-32k-seqlen'\n",
    "      'hyenadna-medium-160k-seqlen'  # inference only on colab\n",
    "      'hyenadna-medium-450k-seqlen'  # inference only on colab\n",
    "      'hyenadna-large-1m-seqlen'  # inference only on colab\n",
    "    '''\n",
    "\n",
    "    # you only need to select which model to use here, we'll do the rest!\n",
    "    pretrained_model_name = 'hyenadna-small-32k-seqlen'\n",
    "\n",
    "    max_lengths = {\n",
    "        'hyenadna-tiny-1k-seqlen': 1024,\n",
    "        'hyenadna-small-32k-seqlen': 32768,\n",
    "        'hyenadna-medium-160k-seqlen': 160000,\n",
    "        'hyenadna-medium-450k-seqlen': 450000,  # T4 up to here\n",
    "        'hyenadna-large-1m-seqlen': 1_000_000,  # only A100 (paid tier)\n",
    "    }\n",
    "\n",
    "    max_length = max_lengths[pretrained_model_name]  # auto selects\n",
    "\n",
    "    # data settings:\n",
    "    use_padding = True\n",
    "    rc_aug = False  # reverse complement augmentation\n",
    "    add_eos = False  # add end of sentence token\n",
    "\n",
    "    # we need these for the decoder head, if using\n",
    "    use_head = False\n",
    "    n_classes = 2  # not used for embeddings only\n",
    "\n",
    "    # you can override with your own backbone config here if you want,\n",
    "    # otherwise we'll load the HF one in None\n",
    "    backbone_cfg = None\n",
    "\n",
    "    device = 'cuda' if torch.cuda.is_available() else 'cpu'\n",
    "    print(\"Using device:\", device)\n",
    "\n",
    "    # instantiate the model (pretrained here)\n",
    "    if pretrained_model_name in ['hyenadna-tiny-1k-seqlen',\n",
    "                                 'hyenadna-small-32k-seqlen',\n",
    "                                 'hyenadna-medium-160k-seqlen',\n",
    "                                 'hyenadna-medium-450k-seqlen',\n",
    "                                 'hyenadna-large-1m-seqlen']:\n",
    "        # use the pretrained Huggingface wrapper instead\n",
    "        model = HyenaDNAPreTrainedModel.from_pretrained(\n",
    "            './checkpoints',\n",
    "            pretrained_model_name,\n",
    "            download=True,\n",
    "            config=backbone_cfg,\n",
    "            device=device,\n",
    "            use_head=use_head,\n",
    "            n_classes=n_classes,\n",
    "        )\n",
    "\n",
    "    # from scratch\n",
    "    elif pretrained_model_name is None:\n",
    "        model = HyenaDNAModel(**backbone_cfg, use_head=use_head, n_classes=n_classes)\n",
    "\n",
    "    # create tokenizer\n",
    "    tokenizer = CharacterTokenizer(\n",
    "        characters=['A', 'C', 'G', 'T', 'N'],  # add DNA characters, N is uncertain\n",
    "        model_max_length=max_length + 2,  # to account for special tokens, like EOS\n",
    "        add_special_tokens=False,  # we handle special tokens elsewhere\n",
    "        padding_side='left', # since HyenaDNA is causal, we pad on the left\n",
    "    )\n",
    "\n",
    "    #### Single embedding example ####\n",
    "\n",
    "    # create a sample 450k long, prepare\n",
    "    sequence = 'ACTG' * int(max_length/4)\n",
    "    tok_seq = tokenizer(sequence)\n",
    "    tok_seq = tok_seq[\"input_ids\"]  # grab ids\n",
    "\n",
    "    # place on device, convert to tensor\n",
    "    tok_seq = torch.LongTensor(tok_seq).unsqueeze(0)  # unsqueeze for batch dim\n",
    "    tok_seq = tok_seq.to(device)\n",
    "\n",
    "    # prep model and forward\n",
    "    model.to(device)\n",
    "    model.eval()\n",
    "    with torch.inference_mode():\n",
    "        embeddings = model(tok_seq)\n",
    "\n",
    "    print(embeddings.shape)  # embeddings here!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "2BYJx_XSiXZr",
    "outputId": "5bc33934-6a5c-4a67-8b96-c6c71db75627"
   },
   "outputs": [],
   "source": [
    "# # uncomment to run! (to get embeddings)\n",
    "inference_single()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "arcIbXNzHmyj"
   },
   "outputs": [],
   "source": [
    "#@title Batch example\n",
    "\"\"\"\n",
    "Let's say you want to do inference on a dataset to grab a lot of embeddings,\n",
    "you can just loop thru a dataloader like this.\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "import json\n",
    "import os\n",
    "import subprocess\n",
    "import transformers\n",
    "from transformers import PreTrainedModel, AutoModelForCausalLM, PretrainedConfig\n",
    "\n",
    "# example, if you want to do look thru a dataloader and get embeddings\n",
    "\n",
    "def infer_loop(model, device, test_loader):\n",
    "    \"\"\"inference loop.\"\"\"\n",
    "    # model.eval()\n",
    "\n",
    "    with torch.inference_mode():\n",
    "        for data, target in test_loader:\n",
    "            data, target = data.to(device), target.to(device)\n",
    "            output = model(data)\n",
    "\n",
    "            # print embedding shape, or do something else here, like save.\n",
    "            print('embeddings shape', output.shape)\n",
    "\n",
    "def inference():\n",
    "\n",
    "    '''\n",
    "    this selects which backbone to use, and grabs weights/ config from HF\n",
    "    4 options:\n",
    "      'hyenadna-tiny-1k-seqlen'   # fine-tune on colab ok\n",
    "      'hyenadna-small-32k-seqlen'\n",
    "      'hyenadna-medium-160k-seqlen'  # inference only on colab\n",
    "      'hyenadna-medium-450k-seqlen'  # inference only on colab\n",
    "      'hyenadna-large-1m-seqlen'  # inference only on colab\n",
    "    '''\n",
    "\n",
    "    # select model\n",
    "    pretrained_model_name = 'hyenadna-medium-160k-seqlen'  # use None if training from scratch\n",
    "\n",
    "    max_lengths = {\n",
    "        'hyenadna-tiny-1k-seqlen': 1024,\n",
    "        'hyenadna-small-32k-seqlen': 32768,\n",
    "        'hyenadna-medium-160k-seqlen': 160000,\n",
    "        'hyenadna-medium-450k-seqlen': 450000,  # T4 up to here\n",
    "        'hyenadna-large-1m-seqlen': 1_000_000,  # only A100 (paid tier)\n",
    "    }\n",
    "\n",
    "    # let's fix the max_length (to reduce the padding amount, conserves memory)\n",
    "    max_length = 500\n",
    "\n",
    "    # data settings:\n",
    "    # we need to choose the dataset and batch size to loop thru\n",
    "    dataset_name = 'human_enhancers_cohn'\n",
    "    batch_size = 4\n",
    "    use_padding = True\n",
    "    rc_aug = False  # reverse complement augmentation\n",
    "    add_eos = False  # add end of sentence token\n",
    "\n",
    "    # we need these for the decoder head, if using\n",
    "    use_head = False\n",
    "    n_classes = 2  # not used for embeddings only\n",
    "\n",
    "    # you can override with your own backbone config here if you want,\n",
    "    # otherwise we'll load the HF one in None\n",
    "    backbone_cfg = None\n",
    "\n",
    "    device = 'cuda' if torch.cuda.is_available() else 'cpu'\n",
    "    print(\"Using device:\", device)\n",
    "\n",
    "    # instantiate the model (pretrained here)\n",
    "    if pretrained_model_name in ['hyenadna-tiny-1k-seqlen',\n",
    "                                 'hyenadna-small-32k-seqlen',\n",
    "                                 'hyenadna-medium-160k-seqlen',\n",
    "                                 'hyenadna-medium-450k-seqlen',\n",
    "                                 'hyenadna-large-1m-seqlen']:\n",
    "        # use the pretrained Huggingface wrapper instead\n",
    "        model = HyenaDNAPreTrainedModel.from_pretrained(\n",
    "            './checkpoints',\n",
    "            pretrained_model_name,\n",
    "            download=True,\n",
    "            config=backbone_cfg,\n",
    "            device=device,\n",
    "            use_head=use_head,\n",
    "            n_classes=n_classes,\n",
    "        )\n",
    "\n",
    "    # from scratch\n",
    "    elif pretrained_model_name is None:\n",
    "        model = HyenaDNAModel(**backbone_cfg, use_head=use_head, n_classes=n_classes)\n",
    "\n",
    "    # create tokenizer\n",
    "    tokenizer = CharacterTokenizer(\n",
    "        characters=['A', 'C', 'G', 'T', 'N'],  # add DNA characters, N is uncertain\n",
    "        model_max_length=max_length + 2,  # to account for special tokens, like EOS\n",
    "        add_special_tokens=False,  # we handle special tokens elsewhere\n",
    "        padding_side='left', # since HyenaDNA is causal, we pad on the left\n",
    "    )\n",
    "\n",
    "    # Sample small dataloader w/ GenomicBenchmarks\n",
    "\n",
    "    ds_test = GenomicBenchmarkDataset(\n",
    "        max_length = max_length,\n",
    "        use_padding = use_padding,\n",
    "        split = 'test',\n",
    "        tokenizer=tokenizer,\n",
    "        dataset_name=dataset_name,\n",
    "        rc_aug=rc_aug,\n",
    "        add_eos=add_eos,\n",
    "    )\n",
    "\n",
    "    ds_test = GenomicBenchmarkDataset(\n",
    "        max_length = max_length,\n",
    "        use_padding = use_padding,\n",
    "        split = 'test',\n",
    "        tokenizer=tokenizer,\n",
    "        dataset_name=dataset_name,\n",
    "        rc_aug=rc_aug,\n",
    "        add_eos=add_eos,\n",
    "    )\n",
    "\n",
    "    test_loader = DataLoader(ds_test, batch_size=batch_size, shuffle=False)\n",
    "\n",
    "    model.to(device)\n",
    "    model.eval()\n",
    "\n",
    "    infer_loop(model, device, test_loader)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "DSkwVPfAHnAG"
   },
   "outputs": [],
   "source": [
    "# uncomment to run!\n",
    "# inference()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "DdWb4ti0iXkB"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "-1rs0t9RIWKt"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "accelerator": "GPU",
  "colab": {
   "collapsed_sections": [
    "5vMJu0wQWUBN"
   ],
   "gpuType": "V100",
   "provenance": [],
   "toc_visible": true
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
